### PR TITLE
Use quoted strings for chunks of code (multiline strings)

### PR DESCRIPTION
**Pull request by MisterDA**
Now that #12514 is merged, we can take advantage of OCaml 4.02's quoted strings, making both the code being emitted, and the formatting of the lines, somewhat less obscure.

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Comment by gasche** [comment]
This looks reasonable and the LaTeX code in particular benefits greatly. I try to find avenues for nitpicks that are *strictly less important* than the topic of the PR itself, and at first I thought that I would be stuck on a PR about lexical conventions. Fear not, we can talk about whitespace!

The use of quoted string literals introduces a dedentation of the content -- we cannot indent to the current indentation level inside the string. This is visually jarring but a reasonable cost to pay for the increased readability of the content. However, I would like to suggest a rule when this happens, that you do not always respect: when string-literal content is going to dedented at the very beginning of a line, I think that the input delimiter that introduces this content should also be dedented at the very beginning of a line.

So this:

```ocaml
  if ctx.has_refill then
    pr ctx {|
let rec __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
                                           _last_action state k =
...
|}
```

should rather be written as

```ocaml
  if ctx.has_refill then
    pr ctx
{|
let rec __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
                                           _last_action state k =
...
|}
```

so that it is clear what is going on. (There are several other instances in this PR.)

**Comment by MisterDA** [comment]
Thanks for the review, I've applied your suggestion.

As a side note, I was tempted to introduce a new function handling the indentation of the generated code in `lex/`, in order to remove the leading `%s` and the trail of `pref pref pref`. My first idea was to scan the code for new lines and add the requested leading spaces, something like:

```ocaml
let pr' ~pref ctx =
  Printf.ksprintf (fun str ->
      let i = ref 0 in
      String.iteri (fun j -> function
          | '\n' when !i = j ->
             Out_channel.output_char ctx.oc '\n';
             incr i;
          | '\n' ->
             Out_channel.output_string ctx.oc pref;
             Out_channel.output_substring ctx.oc str !i (j - !i + 1);
             i := j + 1
          | _ -> ()) str;
      if !i <> String.length str then begin
          Out_channel.output_string ctx.oc pref;
          Out_channel.output_substring ctx.oc str !i (String.length str - !i)
        end)
```

…but maybe that's something more suited for the `Format` module?

File: Changes [hunk 1]
```diff
@@ -189,6 +189,9 @@ _______________
   false positives on s390x
   (Miod Vallat, review by Fabien Buoro)
 
+- #13168: Use quoted strings to clarify code being generated.
```
Comment (ghost) [inline Changes]:
```suggestion
- #13162: Use quoted strings to clarify code being generated.
```

File: lex/outputbis.ml [hunk 2]
```diff
@@ -29,54 +29,54 @@ type ctx = {
...
+          state lexbuf _last_action _buf _len (_curr + 1) _last k
+            (Char.code (Bytes.unsafe_get _buf _curr))
+        else
+          __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
+                                             _last_action
+            state k
+      )
+      lexbuf
+  end
+|}
```
Comment (gasche) [inline lex/outputbis.ml:59]: I think that a newline is missing here -- compared to the previous output. Leaving a blank line at the end looks intentional.

File: lex/outputbis.ml [hunk 3]
```diff
@@ -29,54 +29,54 @@ type ctx = {
...
+    let _last = lexbuf.Lexing.lex_last_pos in
+    let _len = lexbuf.Lexing.lex_buffer_len in
+    let _buf = lexbuf.Lexing.lex_buffer in
+    if _curr < _len then
+      Char.code (Bytes.unsafe_get _buf _curr), _buf, _len,
+                            (_curr + 1), _last
+    else
+      __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
   end
+|}
```
Comment (gasche) [inline lex/outputbis.ml:81]: Same here, I think a newline is missing.

File: ocamldoc/odoc_latex_style.ml [hunk 4]
```diff
@@ -15,79 +15,80 @@
...
-\n     {\\endocamldocindent\\vskip -\\lastskip\
-\n      \\noindent\\quad\\texttt{end}\\medskip}\
-\n\\newenvironment{ocamldocobjectend}\
-\n     {\\noindent\\quad\\texttt{object}\\ocamldocindent}\
-\n     {\\endocamldocindent\\vskip -\\lastskip\
-\n      \\noindent\\quad\\texttt{end}\\medskip}\
-\n\
-\n\\endinput\
-\n"
+let content = {|
```
Comment (gasche) [inline ocamldoc/odoc_latex_style.ml]: I would start with `{|` on the beginning of a new line.

File: ocamldoc/odoc_latex_style.ml [hunk 5]
```diff
@@ -15,79 +15,80 @@
...
+     {\endocamldocindent\vskip -\lastskip
+      \noindent\quad\texttt{end}\medskip}
+\newenvironment{ocamldocobjectend}
+     {\noindent\quad\texttt{object}\ocamldocindent}
+     {\endocamldocindent\vskip -\lastskip
+      \noindent\quad\texttt{end}\medskip}
+
+\endinput
+
+|}
```
Comment (gasche) [inline ocamldoc/odoc_latex_style.ml:93]: Unless I missing something, you are adding a blank line at the end here, which is not intended and not so nice.

**Comment by gasche** [comment]
Your `pr'` function looks reasonable to me. (I don't have the energy to consider and discuss an addition to the Format module in the standard library.) Leaving the code as-is would also be fine, I think.

**Comment by MisterDA** [comment]
Thanks, I've applied your suggestions to keep whitespace identical.

> Your `pr'` function looks reasonable to me. (I don't have the energy to consider and discuss an addition to the Format module in the standard library.) Leaving the code as-is would also be fine, I think.

I'll keep it as-is for now and see later if I can come up with a satisfying design.

**Review by gasche — APPROVED** [review]

**Pull request merged**